src/porepy/numerics/ad/ad_system.py

```diff
@@ -0,0 +1,1866 @@
+
+                # Assure ordering and uniqueness whole equation indexation
+                all_idx = np.unique(np.hstack(img_info.values()))
+
+                # Complementing indices are found by deleting the filtered indices
+                complement_idx = np.delete(all_idx, idx)
+                complement.update({name: complement_idx})
+
+            # If there was no grid-based row filtering, the complement is empty.
+            else:
+                complement.update({name: None})
+        return complement
+
+    def discretize(self, equations: Optional[EquationLike] = None) -> None:
+        """Find and loop over all discretizations in the equation operators, extract
+        unique references and discretize.
+
+        This is more efficient than discretizing on the Operator level, since
+        discretizations which occur more than once in a set of equations will be
+        identified and only discretized once.
+
+        Parameters:
+            equations (optional): a subset of equations.
+                If not provided (None), all known equations will be discretized.
+
+        """
+        equation_names = list(self._parse_equation_like(equations).keys())
+
+        # List containing all discretizations
+        discr: list = []
+        # TODO the search can be done once (in some kind of initialization)
+        for name in equation_names:
+            # this raises a key error if a given equation name is unknown
+            eqn = self._equations[name]
+            # This will expand the list discr with new discretizations.
+            # The list may contain duplicates.
+            discr += self._recursive_discretization_search(eqn, list())
+
+        # Uniquify to save computational time, then discretize.
+        unique_discr = _ad_utils.uniquify_discretization_list(discr)
+        _ad_utils.discretize_from_list(unique_discr, self.mdg)
+
+    def assemble(
+        self,
+        state: Optional[np.ndarray] = None,
+    ) -> tuple[sps.spmatrix, np.ndarray]:
+        """Assemble Jacobian matrix and residual vector of the whole system.
+
+        This is a shallow wrapper of :meth:`assemble_subsystem`, where the subsystem is
+        the complete set of equations, variables and grids.
+
+        Parameters:
+            state (optional): see :meth:`assemble_subsystem`. Defaults to None.
+
+        Returns:
+            sps.spmatrix: Jacobian matrix corresponding to the targeted state.
+                The ordering of the equations (rows) is determined by the order the
+                equations were added. The DOFs (columns) are ordered according the
+                global order.
+            np.ndarray: Residual vector corresponding to the targeted state,
+                scaled with -1 (moved to rhs).
+
+        """
+        return self.assemble_subsystem(state=state)
+
+    def assemble_subsystem(
+        self,
+        equations: Optional[EquationLike] = None,
+        variables: Optional[VariableList] = None,
```

Comment:
Is the subsystem assembly based on cateogories/ tags completely gone or only temporary?
That functionality was intended as a quite useful extension of the EquationManager

Reply:
We certainly need it, but the operation is slightly changed: instead of filtering inside this, and several other, functions, filtering is done on the user side by calling the get-variable function. This reduced list is then used for subsystem assembly. This led to a much simpler treatment of variables inside the functions